### PR TITLE
Refactored code to move all ticking logic for memory requests to Memory.

### PR DESCRIPTION
**Pull request by rahahahat**

### CODE REVIEW

File: src/include/simeng/memory/MMU.hh [hunk 1]
```diff
@@ -66,20 +66,14 @@ class MMU {
...
 
   /** A vector containing all completed Instruction read requests. */
   std::vector<MemoryReadResult> completedInstrReads_;
 
-  /** A queue containing all pending memory requests. */
-  std::queue<FixedLatencyMemoryInterfaceRequest> pendingRequests_;
-
-  /** The number of times this interface has been ticked. */
-  uint64_t tickCounter_ = 0;
+  /** The number of pending requests. */
```
Comment (FinnWilkinson) [inline src/include/simeng/memory/MMU.hh:75]: Change to "pending data requests"

File: src/include/simeng/memory/MemPacket.hh [hunk 2]
```diff
@@ -116,7 +125,9 @@ class MemPacket {
    * 3rd bit indicates whether a MemPacket is faulty (1) or not (0).
    * 4th bit indicates whether a MemPacket should be ignored (1) or not (0).
    * 5th bit indicates whether a MemPacket contains a payload (1) or not (0).
-   * 6th bit indicates whether a MemPacket is an untimed Read (1) or not (0). */
+   * 6th bit indicates whether a MemPacket reads an instruction (1) or not (0).
+   * 7th bit indicates untimed (1) or timed (0) memory access.
```
Comment (FinnWilkinson) [inline src/include/simeng/memory/MemPacket.hh:129]: *indicates whether an untimed (1) or timed (0) memory access is performed."

File: src/include/simeng/memory/FixedLatencyMemory.hh [hunk 3]
```diff
@@ -72,6 +72,22 @@ class FixedLatencyMemory : public Mem {
...
+      handleWriteRequest(req);
+    } else {
+      std::cerr << "[SimEng:FixedLatencyMemory] Invalid MemPacket type for "
+                   "requesting access to memory. Requests to memory should "
+                   "either be of "
+                   "type READ_REQUEST or WRITE_REQUEST."
+                << std::endl;
+      req->markAsFaulty();
+    }
+  }
```
Comment (FinnWilkinson) [inline src/include/simeng/memory/FixedLatencyMemory.hh:90]: Move the body of this function to the main .cc file

File: src/lib/memory/MMU.cc [hunk 4]
```diff
@@ -7,54 +7,33 @@
...
-    } else {
-      bufferRequest(memory::MemPacket::createReadRequest(
-          target.vaddr, target.size, requestId));
-    }
-
-    // Remove the request from the queue
-    pendingRequests_.pop();
-  }
-}
+void MMU::tick() {}
```
Comment (FinnWilkinson) [inline src/lib/memory/MMU.cc:12]: Remove empty function

File: src/lib/memory/MMU.cc [hunk 5]
```diff
@@ -100,9 +79,8 @@ void MMU::setTid(uint64_t tid) { tid_ = tid; }
 std::shared_ptr<Port<std::unique_ptr<MemPacket>>> MMU::initPort() {
   port_ = std::make_shared<Port<std::unique_ptr<MemPacket>>>();
   auto fn = [this](std::unique_ptr<MemPacket> packet) -> void {
-    if (packet->isUntimedRead()) {
-      // Untimed Read only used by instruction requests
-      if (packet->isFaulty()) {
+    if (packet->isInstrRead()) {
+      if (packet->isFaulty() || packet->ignore()) {
         // If faulty, return no data. This signals a data abort.
```
Comment (FinnWilkinson) [inline src/lib/memory/MMU.cc]: Update comment to "If faulty or ignored, ..."

File: src/tools/simeng/main.cc [hunk 6]
```diff
@@ -40,9 +41,12 @@ int simulate(simeng::OS::SimOS& simOS, simeng::Core& core,
     // Tick the core
     core.tick();
 
-    // Tick memory
+    // Tick MMU
```
Comment (FinnWilkinson) [inline src/tools/simeng/main.cc:44]: Remove this as MMU.tick() does nothing

File: test/unit/FixedLatencyMemoryTest.cc [hunk 7]
```diff
@@ -33,11 +33,68 @@ TEST(LatencyMemoryInterfaceTest, FixedWriteData) {
...
+  auto connection =
+      simeng::PortMediator<std::unique_ptr<simeng::memory::MemPacket>>();
+  auto port1 = mmu->initPort();
+  auto port2 = mem->initPort();
+  connection.connect(port1, port2);
+
+  EXPECT_FALSE(mmu->hasPendingRequests());
+
+  // Write a 32-bit value to memory
+  // Should ignore the 7 cycle latency and opt for the interface defined latency
```
Comment (FinnWilkinson) [inline test/unit/FixedLatencyMemoryTest.cc:86]: What 7 cycle latency?

File: test/unit/FixedLatencyMemoryTest.cc [hunk 8]
```diff
@@ -48,9 +105,9 @@ TEST(LatencyMemoryInterfaceTest, FixedWriteData) {
 }
 
 // Test that out-of-bounds memory reads are correctly handled.
-TEST(LatencyMemoryInterfaceTest, UnMappedAddrRead) {
+TEST(LatencyMemoryMemoryTest, UnMappedAddrRead) {
```
Comment (FinnWilkinson) [inline test/unit/FixedLatencyMemoryTest.cc:108]: Change name from `LatencyMemoryMemoryTest`

File: src/include/simeng/memory/MMU.hh [hunk 9]
```diff
@@ -19,7 +19,7 @@ namespace memory {
 
 class MMU {
  public:
-  MMU(const uint16_t latency, VAddrTranslator fn);
+  MMU(VAddrTranslator fn);
```
Comment (FinnWilkinson) [inline src/include/simeng/memory/MMU.hh:23]: I cant comment on the function itself, but given `tick()` now does nothing it can be removed

**Review by FinnWilkinson — CHANGES_REQUESTED** [review]

File: test/unit/FixedLatencyMemoryTest.cc (hunk 7 above)
Comment (rahahahat) [reply test/unit/FixedLatencyMemoryTest.cc:86]: comment from FixedLatencyMemoryInterface tests, I have removed it

File: src/lib/memory/MMU.cc (hunk 4 above)
Comment (rahahahat) [reply src/lib/memory/MMU.cc:12]: resolved offline

File: src/include/simeng/memory/MemPacket.hh (hunk 2 above)
Comment (rahahahat) [reply src/include/simeng/memory/MemPacket.hh:129]: done

File: src/include/simeng/memory/MMU.hh (hunk 1 above)
Comment (rahahahat) [reply src/include/simeng/memory/MMU.hh:75]: done

File: test/unit/FixedLatencyMemoryTest.cc (hunk 8 above)
Comment (rahahahat) [reply test/unit/FixedLatencyMemoryTest.cc:108]: done

File: src/tools/simeng/main.cc [hunk 10]
```diff
@@ -29,7 +30,7 @@ simeng::OS::SimOS simOsFactory(std::shared_ptr<simeng::memory::Mem> memory,
 
 /** Tick the provided core model until it halts. */
 int simulate(simeng::OS::SimOS& simOS, simeng::Core& core,
-             simeng::memory::MMU& mmu) {
+             simeng::memory::MMU& mmu, simeng::memory::Mem& mem) {
```
Comment (FinnWilkinson) [inline src/tools/simeng/main.cc:33]: No point passing MMU reference if it isn't used

Comment (rahahahat) [reply src/tools/simeng/main.cc:33]: MMU is used through `mmu.hasPendingRequests`

Comment (FinnWilkinson) [reply src/tools/simeng/main.cc:33]: I see, my bad

**Pull request merged**